### PR TITLE
fix runlabel functions based on QA feedback

### PR DESCRIPTION
**Pull request by baude**
Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by baude** [comment]
bot, retest this please

File: cmd/podman/runlabel.go [hunk 1]
```diff
@@ -23,7 +26,7 @@ var (
 		},
 		cli.BoolFlag{
 			Name:  "display",
-			Usage: "preview the command that `podman install` would execute",
+			Usage: "preview the command that `runlabel` would execute",
```
Comment (ypu) [inline cmd/podman/runlabel.go]: As display is a bool type flag, maybe we should remove the "`" from the Usage, and didn't add anything behind the flag just like flag quiet.

**Comment by ypu** [comment]
Hi @baude 
Test with the patch. Seems -p will pull image no matter the image is exist or not. This is not consistent with the help text. 

And I have a question about the code. Seems we get opt from env but will make a replacement for image and name. With this implement, we should treat image, name differently with opt when we add labels to image. It is a little wired for me. Is it possible to read them all from env and keep image, name and opts with the same format in the label?

File: docs/podman-container-runlabel.1.md [hunk 2]
```diff
@@ -95,6 +95,23 @@ Require HTTPS and verify certificates when contacting registries (default: true)
...
 tls verification will be used unless the target registry is listed as an insecure registry in registries.conf
 
+## Examples ##
+
+Execute the run label of an image called foobar.
+```
+$ sudo podman container runlabel run foobar
+```
+
+Execute the install label of an image called foobar with additional arguements.
```
Comment (TomSweeneyRedHat) [inline docs/podman-container-runlabel.1.md]: s/arguments/arguements

File: cmd/podman/runlabel.go [hunk 3]
```diff
@@ -187,6 +199,23 @@ func runlabelCmd(c *cli.Context) error {
...
+
+	envmapper := func(k string) string {
+		switch k {
+		case "OPT1":
+			return envmap["OPT1"]
+		case "OPT2":
+			return envmap["OPT2"]
+		case "OPT3":
+			return envmap["OPT3"]
+		}
```
Comment (TomSweeneyRedHat) [inline cmd/podman/runlabel.go]: Might consider a default case here with at least a logrus.debug message.

**Comment by vrothberg** [comment]
I just noticed a problem with optional arguments that will appended to the rulabel command e.g., `podman container runlabel LABEL IMAGE args`. `args` can make some troubles for CLI parsing as it will be interpreted as such at the moment.

File: docs/podman-container-runlabel.1.md (hunk 2 above)
Comment (vrothberg) [reply docs/podman-container-runlabel.1.md]: It's the "run" label below but the comment states "install".

Comment (baude) [reply docs/podman-container-runlabel.1.md]: fixed, ty

File: cmd/podman/runlabel.go [hunk 4]
```diff
@@ -6,9 +6,12 @@ import (
 	"os"
 	"strings"
 
+	"github.com/containers/image/types"
 	"github.com/containers/libpod/cmd/podman/libpodruntime"
 	"github.com/containers/libpod/cmd/podman/shared"
 	"github.com/containers/libpod/libpod/image"
+	image2 "github.com/containers/libpod/libpod/image"
```
[the remark this answers is not in the document]
Comment (baude) [reply cmd/podman/runlabel.go]: artifact of copy and paste; fixed

**Comment by baude** [comment]
> Hi @baude
> Test with the patch. Seems -p will pull image no matter the image is exist or not. This is not consistent with the help text.
> 
> And I have a question about the code. Seems we get opt from env but will make a replacement for image and name. With this implement, we should treat image, name differently with opt when we add labels to image. It is a little wired for me. Is it possible to read them all from env and keep image, name and opts with the same format in the label?

Fixed the logic of -p.  I think the code is behaving as expected? And consistent with atomic cli.  Do you disagree?

File: cmd/podman/runlabel.go (hunk 1 above)
Comment (baude) [reply cmd/podman/runlabel.go]: changed

File: cmd/podman/runlabel.go [hunk 5]
```diff
@@ -23,7 +25,7 @@ var (
 		},
 		cli.BoolFlag{
 			Name:  "display",
-			Usage: "preview the command that `podman install` would execute",
+			Usage: "preview the command the label's command",
```
Comment (mheon) [inline cmd/podman/runlabel.go]: "preview the command the label would run" maybe?

Comment (baude) [reply cmd/podman/runlabel.go]: ok

**Comment by baude** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
I think we should discuss the behaviour of this PR.  @vrothberg and I discussed this in a private IRC, and I think the correct behavior should be:

If basename(command[0]) != docker {
             alias docker=podman
             exec command[0] + args
}
if command[0] exists {
     exec command[0] + args
}
exec /proc/self/exe + args

**Comment by baude** [comment]
> I think we should discuss the behaviour of this PR. @vrothberg and I discussed this in a private IRC, and I think the correct behavior should be:
> 
> If basename(command[0]) != docker {
> alias docker=podman
> exec command[0] + args
> }
> if command[0] exists {
> exec command[0] + args
> }
> exec /proc/self/exe + args

Valentin and I agreed he would submit a PR following mine given it is a change in behavior.  I support that.

**Comment by baude** [comment]
bot, retest this please

**Comment by baude** [comment]
bot, retest this please

**Comment by ypu** [comment]
> Fixed the logic of -p. I think the code is behaving as expected? And consistent with atomic cli. Do you disagree?

No, this is not consistent with atomic cli. In atomic cli the image and name is also work with "$".  Here is the details:

```
#  podman container runlabel  install rhel7/etcd:latest
Command: /proc/self/exe run --rm  --privileged -v /:/host -e HOST=/host -e NAME= -e IMAGE=   /usr/bin/install.sh  
unable to pull : error getting default registries to try: repository name must have at least one component
`/proc/self/exe run --rm  --privileged -v /:/host -e HOST=/host -e NAME= -e IMAGE=   /usr/bin/install.sh  ` failed: exit status 125
```

And here is the install label of the etcd image:
```
            "install": "/usr/bin/docker run --rm $OPT1 --privileged -v /:/host -e HOST=/host -e NAME=$NAME -e IMAGE=$IMAGE $IMAGE $OPT2 /usr/bin/install.sh  $OPT3",
```

**Comment by baude** [comment]
ready for review

File: cmd/podman/shared/funcs.go [hunk 6]
```diff
@@ -24,10 +24,14 @@ func GenerateCommand(command, imageName, name string) []string {
...
 		case "IMAGE=IMAGE":
 			newArg = fmt.Sprintf("IMAGE=%s", imageName)
+		case "IMAGE=$IMAGE":
+			newArg = fmt.Sprintf("IMAGE=%s", imageName)
 		case "NAME":
 			newArg = name
 		case "NAME=NAME":
 			newArg = fmt.Sprintf("NAME=%s", name)
+		case "NAME=$NAME":
+			newArg = fmt.Sprintf("NAME=%s", name)
```
Comment (ypu) [inline cmd/podman/shared/funcs.go]: If we want to treat it like this we should also cover separated "$IMAGE" here. As the atomic cli support it here is the example label:
```
"install": "/usr/bin/docker run --rm $OPT1 --privileged -v /:/host -e HOST=/host -e NAME=$NAME -e IMAGE=$IMAGE $IMAGE $OPT2 /usr/bin/install.sh  $OPT3",
```

Comment (baude) [reply cmd/podman/shared/funcs.go]: i believe that gets taking care of with env substitutions

**Comment by umohnani8** [comment]
LGTM

File: cmd/podman/runlabel.go [hunk 7]
```diff
@@ -154,7 +152,21 @@ func runlabelCmd(c *cli.Context) error {
...
+			}
+			registryCreds = creds
+		}
+		dockerRegistryOptions := image.DockerRegistryOptions{
+			DockerRegistryCreds:         registryCreds,
+			DockerCertPath:              c.String("cert-dir"),
+			DockerInsecureSkipTLSVerify: !c.BoolT("tls-verify"),
+		}
+
+		newImage, err = runtime.ImageRuntime().New(ctx, runlabelImage, c.String("signature-policy"), c.String("authfile"), stdOut, &dockerRegistryOptions, image.SigningOptions{}, false, false)
```
Comment (TomSweeneyRedHat) [inline cmd/podman/runlabel.go]: Probably ought to merge this before the authFile change @QiWang19 is working on.  Pretty sure there'll be a merge dance between the two.

File: cmd/podman/shared/funcs.go (hunk 6 above)
Comment (ypu) [reply cmd/podman/shared/funcs.go]: Hi @baude 
Seems this is not covered bu env. We need fix this:
```
# podman container runlabel -p --tls-verify=false install brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel7/etcd:latest
Command: /proc/self/exe run --rm  --privileged -v /:/host -e HOST=/host -e NAME=brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel7/etcd:latest -e IMAGE=brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel7/etcd:latest   /usr/bin/install.sh  
unable to pull : error getting default registries to try: repository name must have at least one component
`/proc/self/exe run --rm  --privileged -v /:/host -e HOST=/host -e NAME=brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel7/etcd:latest -e IMAGE=brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel7/etcd:latest   /usr/bin/install.sh  ` failed: exit status 125

```

**Comment by vrothberg** [comment]
> Valentin and I agreed he would submit a PR following mine given it is a change in behavior. I support that.

Here's the PR: https://github.com/containers/libpod/pull/1637

**Pull request merged**